Adds additional query syntax for the CDA

floriank opened the pull request:
After having no time for a while these days and thinking a bit about #38, I decided to put in some work about it.

This will add support for more complex queries as noted by @ryansch, in the form of:

```elixir
Entries 
|> content_type("my_content_type") 
|> by(id: "my_id", number_of_warthogs: [lt: 42]) # or similar, no clue what offers the most benefits
|> fetch_all
```
and provide some safeguards around it, e.g. using `by` without a `content_type` call is not permitted, etc.

Feedback, as always, is very welcome.

## TODO

- [x] Entries by `content_type`
- [x] Entries and Assets via `by`
- [x] Array support in `by`, via `[nin]` and `[in]`
- [x] Range support in `by` via `[lte]`, `[gte]`, `[lt]` and `[gt]`
- [x] Existence support
- [x] Full text support via `query` and `[match]`

## Out of scope

- Full text per field
- Geolocation shenanigans
- filtering by mime types

Somebody create some issues for the ones above please :wink: (will probably do tomorrow)

Fixes #38.

floriank committed:
fixes typo in test

floriank committed:
adds mix test watch

for easier tdd

floriank committed:
fixes a small matching bug

floriank committed:
allows for filtering by content type in fetch_all

floriank committed:
adds the ability to filter entries with `by`

floriank committed:
this is not Java

floriank committed:
raises an error in case modifier is incorrect

floriank committed:
thrwos proper errors

floriank committed:
documents `by` for assets and entries

floriank committed:
adds support for assets using by

floriank committed:
moves around some code

Removes configuration into it's own module, also moves request related
functions into it's own `Contentful.Request` module.

floriank committed:
adds support for more modifiers

floriank committed:
adds full_text_search for entries

dlitvakb commented on lib/contentful/request.ex:
```diff
@@ -0,0 +1,131 @@
+  import Contentful.Misc, only: [fallback: 2]
+
+  @agent_header [
+    "User-Agent": "Contentful Elixir SDK"
```
User Agent headers should conform to our `X-Contentful-User-Agent` format, you can find the definition specs here: https://github.com/contentful/contentful.rb/blob/25027ea6cff613af15da27042b88584302c268cf/spec/client_configuration_spec.rb#L223

floriank replied:
Thanks! I think I copied that from an earlier stage of the code.

asaaki replied:
Btw that would be an extra header. I still believe that a standard UA header can still exists, and since Contentful does not care about that one too much, we should be free to set it to a value we like.
Also as a non-official SDK the rules are not that strict anyway.

floriank replied:
Well, I better be prepared :wink:

asaaki commented on lib/contentful/query.ex:
```diff
@@ -266,11 +271,132 @@ defmodule Contentful.Query do
+        |> by(tags: [nin: "maps"])
+        |> fetch_all
+
+  """
```
Maybe let's add some links to the official documentation, too?

floriank replied:
I've added it further above

asaaki reviewed (approved):
Looks quite good. I'm not a query expert though. 😅 
Just tiny nits.

floriank commented on lib/contentful/request.ex:
```diff
@@ -0,0 +1,131 @@
+      |> Keyword.merge(filters)
+      |> URI.encode_query()
+
+    "?#{params}"
```
mmh, good point. Where is my `URLBuilderFactoryBeanProxy` ..

floriank replied:
(Despite the joke, i have to look into the `URL` Utilities here)

floriank replied:
I've refactored for actual URI building via `URI` and `Map.put`, definitely reads better in the tests

floriank committed:
is more explicit about nil values

Co-authored-by: Christoph Grabo <446613+asaaki@users.noreply.github.com>

floriank committed:
adds more explicit fallback for missing API KEY

Co-authored-by: Christoph Grabo <446613+asaaki@users.noreply.github.com>

floriank committed:
adds X-Contentful-User-Agent

floriank committed:
updates request test

..as part of actually writing the officially recognized X-Contentful-User-Agent Header

floriank committed:
refactors into uri construction

as proposed by @asaaki

floriank committed:
addsa link to the docs

floriank committed:
adds changelog entry

asaaki reviewed (approved):
I love it! 😍
